File: apps/loomx-api/src/routes/charts.ts
import { Router, type IRouter } from 'express';
import { ChartsService } from '../services/charts.service';
import { FavoritesService } from '../services/favorites.service';
import { asyncHandler, NotFoundError, ValidationError } from '../middleware/errorHandler';
import { getCurrentUserId } from '../middleware/userContext';
import type { CreateChartDTO, UpdateChartDTO } from '@loomx/types';

const router: IRouter = Router();
const service = new ChartsService();
const favoritesService = new FavoritesService();

router.get('/', asyncHandler(async (req, res) => {
  const charts = await service.list();

  // Prevent caching - list pages should always be fresh
  res.set({
    'Cache-Control': 'no-cache, no-store, must-revalidate',
    'Pragma': 'no-cache',
    'Expires': '0'
  });

  res.json(charts);
}));

router.get('/summary', asyncHandler(async (req, res) => {
  const userId = getCurrentUserId(req);
  const charts = await service.list(userId);

  // Prevent caching - list pages should always be fresh
  res.set({
    'Cache-Control': 'no-cache, no-store, must-revalidate',
    'Pragma': 'no-cache',
    'Expires': '0'
  });

  res.json({ count: charts.length, recent: charts });
}));

router.get('/:id', asyncHandler(async (req, res) => {
  // Disable browser caching - LIVE data only
  res.setHeader('Cache-Control', 'no-cache, no-store, must-revalidate');
  res.setHeader('Pragma', 'no-cache');
  res.setHeader('Expires', '0');

  const chart = await service.getById(req.params.id);
  if (!chart) throw new NotFoundError('Chart');
  res.json(chart);
}));

router.post('/', asyncHandler(async (req, res) => {
  const data: CreateChartDTO = req.body;
  if (!data.name || !data.dataset_id || !data.chart_type) {
    throw new ValidationError('Name, dataset_id, and chart_type are required');
  }
  const userId = getCurrentUserId(req);
  const chart = await service.create(data, userId);
  res.status(201).json(chart);
}));

router.put('/:id', asyncHandler(async (req, res) => {
  const data: UpdateChartDTO = req.body;
  const chart = await service.update(req.params.id, data);
  if (!chart) throw new NotFoundError('Chart');
  res.json(chart);
}));

router.patch('/:id', asyncHandler(async (req, res) => {
  const data: UpdateChartDTO = req.body;
  const chart = await service.update(req.params.id, data);
  if (!chart) throw new NotFoundError('Chart');
  res.json(chart);
}));

router.delete('/:id', asyncHandler(async (req, res) => {
  const deleted = await service.delete(req.params.id);
  if (!deleted) throw new NotFoundError('Chart');
  res.status(204).send();
}));

/**
 * PUT /api/charts/:id/favorite
 * Toggle favorite status for chart
 */
router.put('/:id/favorite', asyncHandler(async (req, res) => {
  const userId = getCurrentUserId(req);
  const { id } = req.params;

  // Get chart to ensure it exists and get its name
  const chart = await service.getById(id);
  if (!chart) {
    throw new NotFoundError('Chart');
  }

  const result = await favoritesService.toggle(
    {
      object_type: 'chart',
      object_id: id,
      object_name: chart.name
    },
    userId
  );

  res.json(result);
}));

export default router;

File: apps/loomx-api/routers/charts.py
"""Charts router — /api/v1/charts."""

from fastapi import APIRouter, Request, Response, HTTPException, Depends
from middleware.auth import require_auth
import services.charts as svc
import services.favorites as fav_svc

router = APIRouter()
NO_CACHE = {
    "Cache-Control": "no-cache, no-store, must-revalidate",
    "Pragma": "no-cache",
    "Expires": "0",
}


@router.get("/charts")
def list_charts(response: Response, user: str = Depends(require_auth)):
    response.headers.update(NO_CACHE)
    return svc.list_charts()


@router.get("/charts/summary")
def charts_summary(response: Response, user: str = Depends(require_auth)):
    response.headers.update(NO_CACHE)
    items = svc.list_charts(user)
    return {"count": len(items), "recent": items}


@router.get("/charts/{chart_id}")
def get_chart(chart_id: str, response: Response, user: str = Depends(require_auth)):
    response.headers.update(NO_CACHE)
    chart = svc.get_chart_by_id(chart_id)
    if not chart:
        raise HTTPException(status_code=404, detail="Chart not found")
    return chart


@router.post("/charts", status_code=201)
def create_chart(data: dict, user: str = Depends(require_auth)):
    if not data.get("name") or not data.get("dataset_id") or not data.get("chart_type"):
        raise HTTPException(status_code=400, detail="Name, dataset_id, and chart_type are required")
    return svc.create_chart(data, user)


@router.put("/charts/{chart_id}")
def update_chart(chart_id: str, data: dict, user: str = Depends(require_auth)):
    result = svc.update_chart(chart_id, data)
    if not result:
        raise HTTPException(status_code=404, detail="Chart not found")
    return result


@router.patch("/charts/{chart_id}")
def patch_chart(chart_id: str, data: dict, user: str = Depends(require_auth)):
    result = svc.update_chart(chart_id, data)
    if not result:
        raise HTTPException(status_code=404, detail="Chart not found")
    return result


@router.delete("/charts/{chart_id}", status_code=204)
def delete_chart(chart_id: str, user: str = Depends(require_auth)):
    deleted = svc.delete_chart(chart_id)
    if not deleted:
        raise HTTPException(status_code=404, detail="Chart not found")


@router.put("/charts/{chart_id}/favorite")
def toggle_chart_favorite(chart_id: str, user: str = Depends(require_auth)):
    chart = svc.get_chart_by_id(chart_id)
    if not chart:
        raise HTTPException(status_code=404, detail="Chart not found")
    return fav_svc.toggle_favorite(
        {"object_type": "chart", "object_id": chart_id, "object_name": chart.get("name")},
        user,
    )
